primary/tests/nodes_bootstrapping_tests.rs

```diff
@@ -0,0 +1,127 @@
+
+    // A cluster of 4 nodes will be created
+    let cluster = Cluster::new(None, None);
+
+    // ==== Start first authority ====
+    cluster.authority(0).start(false, Some(1)).await;
+
+    tokio::time::sleep(node_staggered_delay).await;
+
+    let rounds = authorities_current_round(&cluster).await;
+    assert_eq!(
+        rounds.len(),
+        1,
+        "Expected to have received metric only from one node"
+    );
+    assert!(rounds.values().all(|v| v == &1.0), "We have (f+2) unavailable nodes and expected all nodes to have made progress up to round 1 only");
+
+    // ==== Start second authority ====
+    cluster.authority(1).start(false, Some(1)).await;
+
+    tokio::time::sleep(node_staggered_delay).await;
+
+    let rounds = authorities_current_round(&cluster).await;
+    assert_eq!(
+        rounds.len(),
+        2,
+        "Expected to have received metrics from only two nodes"
+    );
+    assert!(rounds.values().all(|v| v == &1.0), "We have (f+1) unavailable nodes and expected all nodes to have made progress up to round 1 only");
+
+    // ==== Start third authority ====
+    // Now 2f + 1 nodes are becoming available and we expect all the nodes to
+    // start making progress (advance in rounds).
+    cluster.authority(2).start(false, Some(1)).await;
+
+    tokio::time::sleep(node_staggered_delay).await;
+
+    let rounds = authorities_current_round(&cluster).await;
+    assert_eq!(
+        rounds.len(),
+        3,
+        "Expected to have received metrics from only three nodes"
+    );
+    assert!(rounds.values().all(|v| v > &1.0), "We have only (f) unavailable nodes, so all should have made progress more than just the first round");
+
+    // We expect all the nodes to have managed to catch up by now
+    // and be pretty much in similar rounds. The threshold here is
+    // not defined by some rule but more of an expectation.
+    let (min, max) = rounds.values().into_iter().minmax().into_option().unwrap();
+    assert!(max - min <= 2.0, "Nodes shouldn't be that behind");
+
+    // ==== Start fourth authority ====
+    // Now 3f + 1 nodes are becoming available (the whole network) and all the nodes
+    // should make progress
+    cluster.authority(3).start(false, Some(1)).await;
+
+    tokio::time::sleep(node_staggered_delay).await;
+
+    let rounds = authorities_current_round(&cluster).await;
+    assert_eq!(
+        rounds.len(),
+        4,
+        "Expected to have received metrics from all nodes"
+    );
+    assert!(rounds.values().all(|v| v > &1.0), "We have no unavailable nodes, so all should have made progress more than just the first round");
+
+    // We expect all the nodes to have managed to catch up by now
+    // and be pretty much in similar rounds. The threshold here is
+    // not defined by some rule but more of an expectation.
+    let (min, max) = rounds.values().into_iter().minmax().into_option().unwrap();
+    assert!(max - min <= 2.0, "Nodes shouldn't be that behind");
```

Comment:
Same story: what we're hunting for here is consensus reporting sequenced headers.
The `ConsensusListener` (or in the NW code base `Executor`) may be helpful here.